### PR TITLE
CAS prediction, fatty rework, lase changes, minor CAS smoke/flare buff

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request
A few changes to do with CAS. The aim of this PR is to make CAS less bullshit for xenos death-wise, and in particular, more supportive to marines. Because it's cool as fuck to be able to work together with the cool as fuck PO and their cool as fuck jet:

---

### CAS prediction:
-CAS now displays where it will hit, _for marines._ This is a blue display. If you're in it, you're about to be exploded or set on fire or suffer other harm. 
-If you're not in it, you're safe. Keep in mind this may not be 100% accurate if, say, an airlock closes while the ammo's still traveling.

### Lase changes:
-All ammo types (except HV GAU) have their own unique CAS lase (green blinking dot). _Both xenos and marines can see this._
This means, xenos can know what's being fired at them (and direction for GAU/laser). They don't know the AoE or time until it hits.

### CAS flare buff:
Flare actually hits where you want it to. Previously it would hit anywhere within a 4 tile radius.
Flare doesn't murder marines standing near it.
Effect of the actual flare itself is unchanged. But hopefully this means flare isn't a % chance for you to fuck over your own team anymore, for using a support weapon.
Also removed explosion from smoke minirockets (not tfoot), because you can't even use them for bluffing anymore due to lase changes

### Fatty rework:
Okay, so this might be the change everybody screams at me for the most.
Instead of its old function (initial small explosion, 8 delayed explosions nearby), it now deletes resin walls nearby. _This does not harm marines._
I'll explain why this is good further down

### CAS prediction:
This allows CAS to support marines easier, and lets marines know before they are about to be absolutely obliterated by their own teammate; encouraging teamwork on both sides.

### Lase changes:
I know this removes a lot of bluffing potential for CAS, and as such a _really fucking big part of CAS_ is just gone. It's still a good change. This isn't fun at all for xenos, and can completely ruin the game for them. It needs to be removed, and to compensate there's other stuff to help marines on the ground and ways to shoot xenos more often. Bullshit deaths just aren't fun.

### CAS flare buff:
-Flare and smoke are already regarded as garbage
-You can no longer make xenos think they will be instakilled by them.
-They are now completely useless
-By making them more reliable and usable within large clusters of marines, they should actually be fucking viable for the first time in a thousand years

### Fatty changes:
To start, this won't be a maze clearer. It costs 200 CAS points, and clears an area 5 tiles wide. For wider maze destruction, this gets **ludicrously** expensive. However, there is currently nothing else in the game for marines that clears enough walls to move, while not killing your team in the process (like explosions tend to do). If this is too OP, radius and cost can be adjusted. If it's not good enough, vice versa, or could kill weeds and structures too. The key thing is that it's another option (out of the current 2, flares and tanglefoot spam) that lets CAS help marines in tight spaces. Again, without killing them in the process.

The current state of fatty, as I've been told, isn't really the best. If it hits to kill, xenos often bounce around the edges. If you want to clear space with it, the explosions don't really connect well enough to raze structures. If we buff fatty, all I see is it either becoming Murderer of Xenos, because the only way you really _can_ balance it is by increasing the damage.

Also code-wise calculating prediction for 9 different explosions would be painful so we're not doing that. Fatties already freeze the game for a couple seconds on their own

## Changelog
:cl: breadhunt, thank Elsa and Ciel for sprites
add: Marine CAS prediction matrix. A prediction for where CAS ammo will hit, before it arrives, now shows up on marine HUD.
imageadd: Various new CAS lase sprites, for:
add: CAS lase (flashing green dot) now has a sprite for each type of ammo fired (except HV GAU). Both xenos and marines see this.
balance: This means both marines *and* xenos can see what ammo's coming. Sorry, POs.
balance: Removed deviation from CAS flares
balance: Removed explosion from CAS flares and CAS smoke. Marines are finally safe
expansion: Complete fatty rework.
balance: Instead of a cluster explosion, fatty now destroys all walls in a 5 tile radius around impact.
balance: This is non-lethal; feel free to use it around brazilled marines. Point cost means you can't really spam it, however.
/:cl:

### CODE REVIEW

File: code/game/objects/effects/overlays.dm [hunk 1]
```diff
@@ -49,14 +49,92 @@
...
+//Minirockets
+/obj/effect/overlay/blinking_laser/tfoot
+	icon_state = "tanglefoot_target"
+
+/obj/effect/overlay/blinking_laser/smoke
+	icon_state = "smoke_target"
+
+/obj/effect/overlay/blinking_laser/flare
+	icon_state = "flare_target"
+	color = "#881243" //todo
```
Comment (TiviPlus) [inline code/game/objects/effects/overlays.dm:74]: todo

File: code/game/objects/effects/overlays.dm [hunk 2]
```diff
@@ -49,14 +49,92 @@
...
+
+//Prediction lines. Those horizontal blue lines you see when CAS fires something
+/obj/effect/overlay/blinking_laser/marine/lines
+	layer = WALL_OBJ_LAYER //Above walls/items, not above mobs
+	icon_state_on = "middle"
+
+/obj/effect/overlay/blinking_laser/marine/lines/Initialize()
+	. = ..()
+	dir = pick(1,2,4,8) //Randomises type, for variation
+
```
Comment (TiviPlus) [inline code/game/objects/effects/overlays.dm:137]: use the direction defines

File: code/game/objects/effects/overlays.dm [hunk 3]
```diff
@@ -135,7 +213,7 @@
 /obj/effect/overlay/temp/laser_target/cas
 	icon_state = "laser_target_coordinate"
 
-/obj/effect/overlay/temp/laser_target/cas/Initialize(mapload, effect_duration, named, assigned_squad = null)
+/obj/effect/overlay/temp/laser_target/cas/Initialize(mapload, named, assigned_squad = null)
```
Comment (TiviPlus) [inline code/game/objects/effects/overlays.dm:216]: explain this

File: code/game/objects/effects/overlays.dm [hunk 4]
```diff
@@ -260,10 +338,3 @@
 	alpha = 0
 	vis_flags = NONE
 	blocks_emissive = NONE
-
-/obj/effect/overlay/temp/timestop_effect
-	icon = 'icons/effects/160x160.dmi'
-	icon_state = "time"
-	pixel_x = -60
-	pixel_y = -50
-	alpha = 70
```
Comment (TiviPlus) [inline code/game/objects/effects/overlays.dm:269]: ???

File: code/game/objects/structures/dropship_ammo.dm [hunk 5]
```diff
@@ -28,6 +28,16 @@
 	///Type of ammo
 	var/ammo_type
 
+	///Explosion ratings, for explosive weapons
+	var/devastating_explosion_range = 0 //HE OB uses this
+	var/heavy_explosion_range = 0 //Mortar splash damage uses this
+	var/light_explosion_range = 0 //HEDP uses this
+	///Fire radius, for incendiary weapons
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:35]: A) autodoc
B) I dont know what x uses this is supposed to men please make useful comments or remove them

File: code/game/objects/structures/dropship_ammo.dm [hunk 6]
```diff
@@ -70,6 +80,129 @@
...
+/obj/structure/ship_ammo/proc/get_turfs_to_impact(turf/epicenter, attackdir = NORTH)
+	switch(prediction_type)
+		if(CAS_AMMO_EXPLOSIVE)
+			return get_explosion_impact(epicenter)
+		if(CAS_AMMO_INCENDIARY)
+			return filled_turfs(epicenter, fire_range, "circle")
+
+	//If it's CAS_AMMO_HARMLESS, we don't need to do anything
+
+	return FALSE //For anything else needed, add a special version of this proc in the subtype
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:95]: this does not return booleans

File: code/game/objects/structures/dropship_ammo.dm [hunk 7]
```diff
@@ -70,6 +80,129 @@
...
+		var/turf/affected_turf = t
+		var/dist = turfs_in_range[epicenter]
+		var/turf/expansion_wave_loc = epicenter
+
+		do
+			var/expansion_dir = get_dir(expansion_wave_loc, affected_turf)
+			if(ISDIAGONALDIR(expansion_dir)) //If diagonal we'll try to choose the easy path, even if it might be longer. Damn, we're lazy.
+				var/turf/step_NS = get_step(expansion_wave_loc, expansion_dir & (NORTH|SOUTH))
+				if(!turfs_in_range[step_NS])
+					current_exp_block = step_NS.density ? step_NS.explosion_block : 0
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:145]: is this all copy paste

File: code/game/objects/structures/dropship_ammo.dm [hunk 8]
```diff
@@ -70,6 +80,129 @@
...
+	turfs_in_range[epicenter] = current_exp_block
+
+	var/list/turfs_impacted = list(epicenter)
+	var/list/outline_turfs_impacted = list()
+
+	for(var/t in turfs_in_range)
+		if(!isnull(turfs_by_dist[t])) //Already processed.
+			continue
+
+		var/turf/affected_turf = t
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:136]: use AS

File: code/game/objects/structures/dropship_ammo.dm [hunk 9]
```diff
@@ -264,48 +416,50 @@
...
 
 /obj/structure/ship_ammo/rocket/napalm/detonate_on(turf/impact, attackdir = NORTH)
 	impact.ceiling_debris_check(3)
-	explosion(impact, 2, 3, 4, 6, small_animation = TRUE) //relatively weak
-	flame_radius(5, impact, 60, 30) //cooking for a long time
+	explosion(impact, devastating_explosion_range, heavy_explosion_range, light_explosion_range, small_animation = TRUE) //relatively weak
+	flame_radius(fire_range, impact, 60, 30) //cooking for a long time
 	var/datum/effect_system/smoke_spread/phosphorus/warcrime = new
-	warcrime.set_up(6, impact, 7)
+	warcrime.set_up(fire_range + 1, impact, 7) //we want this a bit bigger than the fire
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:462]: useless comments

File: code/game/objects/structures/dropship_ammo.dm [hunk 10]
```diff
@@ -397,16 +568,20 @@
...
+	heavy_explosion_range = 0
+	light_explosion_range = 2
+	prediction_type = CAS_AMMO_EXPLOSIVE
+	cas_effect = /obj/effect/overlay/blinking_laser/flare
 
 /obj/structure/ship_ammo/minirocket/illumination/detonate_on(turf/impact, attackdir = NORTH)
 	impact.ceiling_debris_check(2)
-	var/turf/offset_impact = pick(range(3, impact))
-	explosion(offset_impact, 0, 0, 2, 2, throw_range = 0)// Smaller explosion to prevent this becoming the PO meta
+	explosion(impact, devastating_explosion_range, heavy_explosion_range, light_explosion_range, throw_range = 0)// Smaller explosion to prevent this becoming the PO meta
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_ammo.dm:579]: useless comments

File: code/game/objects/structures/dropship_equipment.dm [hunk 11]
```diff
@@ -628,9 +628,25 @@
...
 		playsound(target_turf, ammo_warn_sound, 70, 1)
-	var/obj/effect/overlay/blinking_laser/laser = new (target_turf)
+
+	//Lase
+	var/obj/effect/overlay/blinking_laser/laserdot = new SA.cas_effect(target_turf)
+	laserdot.dir = attackdir
+	var/list/effects_to_delete = list(laserdot) //Let's make a list of things we'll need to delete on impact
+
+	//Marine-only visuals
+	var/predicted_dangerous_turfs = SA.get_turfs_to_impact(target_turf, attackdir) //Gets a list of turfs that could be dangerous
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_equipment.dm:638]: useless comments

File: code/game/objects/structures/dropship_equipment.dm [hunk 12]
```diff
@@ -628,9 +628,25 @@
...
+		effects_to_delete += new /obj/effect/overlay/blinking_laser/marine/lines(impact)
+
 	addtimer(CALLBACK(SA, /obj/structure/ship_ammo.proc/detonate_on, target_turf, attackdir), ammo_travelling_time)
-	addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, laser), ammo_travelling_time)
+	addtimer(CALLBACK(src, .proc/cleanup_effects, effects_to_delete), ammo_travelling_time)
+
+/// Qdels everything in effects_to_delete
+/obj/structure/dropship_equipment/weapon/proc/cleanup_effects(effects_to_delete)
+	for(var/obj/effect/effect in effects_to_delete)
+		qdel(effect)
```
Comment (TiviPlus) [inline code/game/objects/structures/dropship_equipment.dm:649]: why does this not use QDEL_LIST()

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by Kuro020 — APPROVED** [review]
should probably post a video demonstrating it, but sprites are okay

**Comment by breadhunt** [comment]
i'll post videos + review compliance tomorrow, the ones up on discord are a bit outdated so i'll have to re-record them

File: code/game/objects/effects/overlays.dm [hunk 13]
```diff
@@ -49,14 +49,92 @@
...
+
+/obj/effect/overlay/blinking_laser/fatty
+	icon_state = "fatty_target"
+
+/obj/effect/overlay/blinking_laser/napalm
+	icon_state = "napalm_target"
+
+//Marine-only visuals. Prediction HUD, etc. Does not show without marine headset
+/obj/effect/overlay/blinking_laser/marine
+	name = "Prediction matrix"
```
Comment (XSlayer300) [inline code/game/objects/effects/overlays.dm:107]:
```suggestion
	name = "prediction matrix"
```

**Comment by Lumipharon** [comment]
So does your fatty rework means you drop a bomb that destroys walls but doesn't hurt mobs?

**Comment by TiviPlus** [comment]
Yea is there a vid of the new fatty

File: code/game/objects/effects/overlays.dm (hunk 3 above)
Comment (breadhunt) [reply code/game/objects/effects/overlays.dm]: I fucked up local. Same for wraith, should be fixed

**Comment by breadhunt** [comment]
https://user-images.githubusercontent.com/80229419/158451200-7c468981-f542-4fd8-90fc-38c65768f404.mp4

If there's anything to improve visual-wise (or otherwise), please let me know
Video was taken before I actually remembered to remove the flare explosion. Please ignore it, it's not there anymore

**Comment by private-tristan** [comment]
this is amazing, especially how you can see CAS as blue dots  as a xeno so when I play defiler and emit green smoke I won't die to nearly invisible green dot on top of green smoke :D

**Comment by novaepee** [comment]
> can xenos see cas as anything other than the blinking green dots? it's a bit hard to see, especially when in things like defiler smoke

No. Only dot.

**Comment by Lumipharon** [comment]
So what is the fatty actually supposed to be/what does it do? It's a huge ass missile/bomb but it doesn't actually do damage, and just deletes resin walls in a relatively small radius. Space magic, man,

It also seems like a horribly inefficient way of removing free walls with a 200 point weapon.

**Comment by Al-1ce** [comment]
> can xenos see cas as anything other than the blinking green dots? it's a bit hard to see, especially when in things like defiler smoke

Defiler smoke I believe is intentionally hard to see through. Although I'm sure somebody could find a way to make it so that it is transparent for xenos exclusively.

**Comment by breadhunt** [comment]
> So what is the fatty actually supposed to be/what does it do? It's a huge ass missile/bomb but it doesn't actually do damage, and just deletes resin walls in a relatively small radius. Space magic, man,
> 
> It also seems like a horribly inefficient way of removing free walls with a 200 point weapon.

It's supposed to clear walls out, instantly, without risk of blocking or killing marines. This can be used to get people out of bad situations, pushing into a flank, removing cover, connecting two fronts etc etc etc etc etc - all of which an explosion wouldn't do effectively and/or would blow up or block marines in the process.

I'm not thinking of it as Plasmacutter 2, it creates space instantly when it's needed *now*. I think that it needs to be expensive or it'll just be spammed and become a primary maze clearer. Which would be absolutely horrid to play against

As for range, I don't think it's too small, honestly. 5 tiles is almost half a screen, which is enough to evade, dodge, whatever. If it does need buffing that's not hard to do, as said in the PR cost/range/additional deatures can be added



> can xenos see cas as anything other than the blinking green dots? it's a bit hard to see, especially when in things like defiler smoke

Xenos just see dots yeah. With Elsa's epic sprites they should be a bit easier to notice though

**Comment by Lumipharon** [comment]
I know what it's doing mechanics wise, I mean what is it doing fluff wise? It seems very gamey to just specificly delete resin walls and do nothing else.

**Comment by Araulius** [comment]
> I know what it's doing mechanics wise, I mean what is it doing fluff wise? It seems very gamey to just specificly delete resin walls and do nothing else.

it should delete all structures in a radius. metal included.

**Comment by breadhunt** [comment]
> I know what it's doing mechanics wise, I mean what is it doing fluff wise? It seems very gamey to just specificly delete resin walls and do nothing else.

Feel free to suggest ideas, I can't think of any



> > I know what it's doing mechanics wise, I mean what is it doing fluff wise? It seems very gamey to just specificly delete resin walls and do nothing else.
> 
> it should delete all structures in a radius. metal included.

As mentioned in the PR we can buff it by making it destroy structures. Not metal though, Keeper's for that

**Comment by krutoi2000** [comment]
If you're updating CAS, can you make shortcuts for his weaponry slots?

**Comment by breadhunt** [comment]
conflict should be resolved i think


> If you're updating CAS, can you make shortcuts for his weaponry slots?

i'm not smart enough for this, sorry

**Pull request merged**